CONTRIBUTING.md

```diff
@@ -0,0 +1,602 @@
+    with 2 blank lines. The first / last statement of a namespace is separated
+    by two blank lines from the opening / closing brace of the namespace.
+    1.  _exception_: if the first __or__ the last statement in the namespace is
+    another namespace, then no blank lines are required
+        _example_:
+        ```c++
+        namespace foo {
+
+
+        struct x {
+        };
+
+
+        }  // namespace foo
+
+
+        namespace bar {
+        namespace baz {
+
+
+        void f();
+
+
+        }  // namespace baz
+        }  // namespace bar
+        ```
+
+    2.  _exception_: in header files whose only purpose is to _declare_ a bunch
+        of functions (e.g. the `*_kernel.hpp` files) these declarations can be
+        separated by only 1 blank line (note: standard rules apply for all other
+        statements that might be present in that file)
+    3.  _exception_: "related" statement can have 1 blank line between them.
+        "Related" is not a strictly defined adjective in this sense, but is in
+        general one of:
+
+        1.  overload of a same function,
+        2.  function / class template and it's specializations,
+        3.  macro that modifies the meaning or adds functionality to the
+            previous / following statement.
+
+        However, simply calling function `f` from function `g` does not imply
+        that `f` and `g` are "related".
+2.  Statements within structures / classes are separated with 1 blank line.
+    There are no blank lines betweeen the first / last statement in the
+    structure / class.
+    1.  _exception_: there is no blank line between an access modifier (`private`, `protected`, `public`) and the following statement.
+       _example_:
+        ```c++
+        class foo {
+        public:
+            int get_x() const noexcept { return x_; }
+
+            int &get_x() noexcept { return x_; }
+
+        private:
+            int x_;
+        };
+        ```
+
+3.  Function bodies cannot have multiple consecutive blank lines, and a single
+    blank line can only appear between two logical sections of the function.
+4. Unit tests should follow the [AAA](http://wiki.c2.com/?ArrangeActAssert)
+   pattern, and a single blank line must appear between consecutive "A"
+   sections. No other blank lines are allowed in unit tests.
+5.  Enumeration definitions should have no blank lines between consecutive
+    enumerators.
+
+
+### Include statement grouping
+
+In general, all include statements should be present on the top of the file,
+ordered in the following groups, with two blank lines between each group:
+
+1. Related header file (e.g. `core/foo/bar.hpp` included in `core/foo/bar.cpp`,
+   or in the unit test`core/test/foo/bar.cpp`)
+2. Standard library headers (e.g. `vector`)
+3. Executor specific library headers (e.g. `omp.h`)
+4. System third-party library headers (e.g. `papi.h`)
+5. Local third-party library headers
+6. Public Ginkgo headers
+7. Private Ginkgo headers
```

Comment:
I would maybe add a precision (I don't know if that's the correct place), that there should be no private ginkgo headers in any public ginkgo header file.